icaruscode/TPC/Tracking/ICARUSPandora/LArPandoraHitCollectionToolICARUS_tool.cc

```diff
@@ -0,0 +1,56 @@
+/**
+ *  @file  larpandora/LArPandoraInterface/Tools/LArPandoraHitCollectionToolICARUS_tool.cc
+ * 
+ *  @brief Define class for hit collection tools and implements ICARUS tool, with a duplication filter
+ * 
+ */
+
+#include "art/Utilities/ToolMacros.h"
+
+#include "larpandora/LArPandoraInterface/LArPandoraHitCollectionTool.h"
+
+#include "art/Framework/Principal/Event.h"
+#include "lardataobj/RecoBase/Hit.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+#include <map>
+
+namespace lar_pandora {
+
+  class LArPandoraHitCollectionToolICARUS : public IHitCollectionTool
+  {
+  public:
+    explicit LArPandoraHitCollectionToolICARUS(const fhicl::ParameterSet& pset);
+    void CollectHits(const art::Event& evt, const std::string& label, HitVector& hitVector) override;
+  };
+
+  LArPandoraHitCollectionToolICARUS::LArPandoraHitCollectionToolICARUS(const fhicl::ParameterSet& pset){}
+
+  void LArPandoraHitCollectionToolICARUS::CollectHits(const art::Event& evt, const std::string& label, HitVector& hitVector)
+  {
+    std::map< std::pair<raw::ChannelID_t, int>, unsigned int > channelTimeHits; // converting the float PeakTime to int for the second item in pair
+
+    art::Handle<std::vector<recob::Hit>> theHits;
+    evt.getByLabel(label, theHits);
+
+    if (!theHits.isValid()) {
+      mf::LogDebug("LArPandoraHitCollectionToolICARUS") << "  Failed to find hits... " << std::endl;
+      return;
+    }
+    else {
+      mf::LogDebug("LArPandoraHitCollectionToolICARUS") << "  Found: " << theHits->size() << " Hits " << std::endl;
+    }
+
+    for (unsigned int i = 0; i < theHits->size(); ++i) {
+      const art::Ptr<recob::Hit> hit(theHits, i);
+      if ( channelTimeHits.find(std::make_pair(hit->Channel(), int(hit->PeakTime()))) == channelTimeHits.end() ) channelTimeHits[std::make_pair(hit->Channel(), int(hit->PeakTime()))] = 1;
+      else continue;
```

Comment:
A couple of tricks to avoid double map lookup (`find()` then `operator[]`), both relying on the fact that map `operator[]` creates an element if it does not exist, and it assigns the "default" value (`0` for integers):
* one liner:
    ```suggestion
          if ( std::exchange(channelTimeHits[std::make_pair(hit->Channel(), int(hit->PeakTime()))], 1) ) continue;
    ```
    This is a decently common pattern in "atomic" operations (operations which must not be interrupted e.g. by other threads), but it is not particularly readable.
* if that's to cryptic, alternative:
    ```suggestion
          int& count = channelTimeHits[std::make_pair(hit->Channel(), int(hit->PeakTime()))];
          if (count) continue;
          else count = 1;
    ```

Reply:
Thanks for the suggestions -- I will try to add one of these and test the behavior on a test event.

Reply:
Addressed.